corehq/apps/saved_reports/models.py

```diff
@@ -127,18 +127,48 @@ def by_domain_and_owner(cls, domain, owner_id, report_slug=None,
         if limit is not None:
             kwargs['limit'] = limit
 
-        result = cache_core.cached_view(
-            db,
-            "reportconfig/configs_by_domain",
-            reduce=False,
-            include_docs=True,
-            startkey=key,
-            endkey=key + [{}],
-            wrapper=cls.wrap,
-            **kwargs
-        )
+        if not include_shared:
+            result = cache_core.cached_view(
+                db,
+                "reportconfig/configs_by_domain",
+                reduce=False,
+                include_docs=True,
+                startkey=key,
+                endkey=key + [{}],
+                wrapper=cls.wrap,
+                **kwargs
+            )
+        else:
+            # Get user's own configs on domain
+            user_configs = cache_core.cached_view(
+                db,
+                "reportconfig/configs_by_domain",
+                reduce=False,
+                include_docs=False,
+                startkey=key,
+                endkey=key + [{}],
+                wrapper=cls.wrap,
+                **kwargs
+            )
+            user_configs = [c['id'] for c in user_configs]
+            # Include all shared configs
+            user_configs.extend(cls.shared_on_domain(domain, only_id=True))
+
+            result = [ReportConfig.get(id) for id in set(user_configs)]
+
         return result
 
+    @classmethod
+    def shared_on_domain(cls, domain, only_id=False, **kwargs):
+        shared_config_ids = []
+        for rn in ReportNotification.by_domain(domain, **kwargs):
+            shared_config_ids.extend(rn.config_ids)
```

Comment:
Both times that you use `shared_config_ids` you convert it to a set. You could create it as a set like so:

```suggestion
        shared_config_ids = {
            id_ for rn in ReportNotification.by_domain(domain, **kwargs)
            for id_ in rn.config_ids
        }
```